node/network/bridge/src/lib.rs

```diff
@@ -0,0 +1,999 @@
+
+	Ok(Some(NetworkBridgeEvent::OurViewChange(local_view.clone())))
+}
+
+async fn run_network<N: Network>(
+	mut net: N,
+	mut ctx: impl SubsystemContext<Message=NetworkBridgeMessage>,
+	action_inspector: impl Fn(&Action), // side-channel for tests to inspect internals
+) -> SubsystemResult<()> {
+	let mut event_stream = net.event_stream().fuse();
+
+	// Most recent heads are at the back.
+	let mut live_heads = Vec::with_capacity(MAX_VIEW_HEADS);
+	let mut local_view = View(Vec::new());
+
+	let mut peers: HashMap<PeerId, PeerData> = HashMap::new();
+	let mut event_producers = BTreeMap::new();
+
+	loop {
+		let action = {
+			let subsystem_next = ctx.recv().fuse();
+			let mut net_event_next = event_stream.next().fuse();
+			futures::pin_mut!(subsystem_next);
+
+			let action = futures::select! {
+				subsystem_msg = subsystem_next => Some(action_from_overseer_message(subsystem_msg)),
+				net_event = net_event_next => action_from_network_message(net_event),
+			};
+
+			match action {
+				Some(a) => a,
+				None => continue,
+			}
+		};
+
+		action_inspector(&action);
+
+		match action {
+			Action::RegisterEventProducer(protocol_id, event_producer) => {
+				// insert only if none present.
+				if let BEntry::Vacant(entry) = event_producers.entry(protocol_id) {
+					let event_producer = entry.insert(event_producer);
+
+					// send the event producer information on all connected peers.
+					let mut messages = Vec::with_capacity(peers.len() * 2);
+					for (peer, data) in &peers {
+						messages.push(event_producer(
+							NetworkBridgeEvent::PeerConnected(peer.clone(), data.role.clone())
+						));
+
+						messages.push(event_producer(
+							NetworkBridgeEvent::PeerViewChange(peer.clone(), data.view.clone())
+						));
+					}
+
+					ctx.send_messages(messages).await?;
+				}
+			}
+			Action::SendMessage(peers, protocol, message) => {
+				let mut message_producer = stream::iter({
+					let n_peers = peers.len();
+					let mut message = Some(
+						WireMessage::ProtocolMessage(protocol, message).encode()
+					);
+
+					peers.iter().cloned().enumerate().map(move |(i, peer)| {
+						// optimization: avoid cloning the message for the last peer in the
+						// list. The message payload can be quite large. If the underlying
+						// network used `Bytes` this would not be necessary.
+						let message = if i == n_peers - 1 {
+							message.take()
+								.expect("Only taken in last iteration of loop, never afterwards; qed")
+						} else {
+							message.as_ref()
+								.expect("Only taken in last iteratio of loop, we are not there yet; qed")
+								.clone()
+						};
+
+						Ok(NetworkAction::WriteNotification(peer, message))
+					})
+				});
+
+				net.action_sink().send_all(&mut message_producer).await?;
+			}
+			Action::ReportPeer(peer, rep) => {
+				// let binding required here to avoid unnecessary sync bound.
+				let report_fut = net.report_peer(peer, rep);
+				report_fut.await?
```

Comment:
```suggestion
				net.report_peer(peer, rep).await?;
```

Now that the methods on `Network` take a mutable reference to `self` you can call `await` on the `Future` directly.